### ISSUE TITLE
Should SharpGLTF throw an exception when an unsupported extension is required?

### ISSUE THREAD

**Issue by BoyBaykiller** [vpenades/SharpGLTF#237]
Currently SharpGLTF throws when an unsupported extension is required by a glTF file:
https://github.com/vpenades/SharpGLTF/blob/b775097d57ad10db51d1ac2030e1117de220afe7/src/SharpGLTF.Core/Schema2/gltf.ExtensionsFactory.cs#L303

Personally I think this is a bit of an overreaction. What if the user wants to process the extension themselves. What if it's a custom extension that will never be supported by SharpGLTF. glTF-Validator doesn't care if I specify an unknown required extension.

What do you think of this?

**Comment by vpenades** [comment]
glTF specification distinguishes between optional and required extensions.

An optional extension is an extension that usually contains metadata or adds additional drawing features that do not change the behavior of the schema.

But extensions can also be tagged as required, which means that the entire model cannot be evaluated unless the given extension is taken into account.

gltf validator is a somewhat special case, because it just gives warnings on unknown extensions because otherwise it would be unusable. After all, gltf validator just checks the model and does not render it.

SharpGLTF handles that internally, and so far I haven't exposed an API to handle used/required.

See https://github.com/vpenades/SharpGLTF/blob/b775097d57ad10db51d1ac2030e1117de220afe7/src/SharpGLTF.Core/Schema2/gltf.ExtensionsFactory.cs#L249

**Comment by vpenades** [comment]
whether an extension is declared as used or required is really up to whether the model is correctly renderable on an engine unable to understand that extensions, but also to prevent the engine to crash with unrecognizable data, so how would you signal that the the engine that loads the model?

The rule of thumb is that if it is safe for an engine to load your model without understanding your extension, it's probably fine to declare the extension as used.

But if your extension changes things in a way that the engine would crash, or render the model so wrong that it would be unrecognizable, then it's when you would use the required.

There's a tweak in the library, you can remove the validation at load time with one of the load extra parameters... but it disables -all the validation- which means you're in your own if there's issues.

**Comment by BoyBaykiller** [comment]
Ok. I've tried `new ReadSettings() { Validation = SharpGLTF.Validation.ValidationMode.Skip }`, but the exception is still thrown.

**Comment by vpenades** [comment]
hmm... i'll take that into account, that validation is somewhat special

**Comment by BoyBaykiller** [comment]
Blender puts `KHR_materials_variants` in required so I can't load any models outputted by blender using this extension. I still think SharpGLTF shouldn't refuse to load the model when encounting unsupported required extension.

**Comment by vpenades** [comment]
Do these models actually use the extension? Or blender exporter blindly adds the extension requirement even if it's not used?

**Comment by BoyBaykiller** [comment]
Yes, the model defines two variants. Use this [Dragon Attenuation model](https://github.com/KhronosGroup/glTF-Sample-Assets/tree/main/Models/DragonAttenuation/glTF) to reproduce. It has `KHR_materials_variants` only in `extensionsUsed` but after exporting with blender it's in `extensionRequired`.

**Comment by vpenades** [comment]
I've been trying to find information and I can't find any information that KHR_materials_variants should be set as "required".

>  I still think SharpGLTF shouldn't refuse to load the model when encounting unsupported required extension.

I understand that in your case it could be fine to load these models in this specific use case, but the glTF specification has "used" and "required" specifically to signal when you are safe to load the model and when not. So SharpGLTF is following the specification of refusing to load a model when it does not support an extension the model is signaling as "required". You might think that these models are safe to load anyway, but the library cannot know in advanced when the model is "lying" about it, there's many required extensions that are really required and would crash or produce incorrect results downstream if loaded.

_It's responsability of the writer library to correctly signal what's used and what's required._

As I said, I could consider removing the check when the Skip flag is set, but you must understand that what you're asking me is to break a feature that is working correctly in my library, because some other library is not following the standard correctly.

Is there an open issue in the blender exporter library requesting to save Variants as used and not as required?

**Comment by BoyBaykiller** [comment]
I agree with everything you have said. The reason I would not let SharpGLTF throw an exception for unsupported extensions is not because I think that it's more correct, but as a form of leniency to avoid cases like this. 

I've opended a issue regarding KHR_materials_variants on the blender project.

**Comment by vpenades** [comment]
Reviewing the issue, there's an additional problem about allowing glTFs with unsupported but required extensions to load.

When SharpGLTF encounters an unsupported extension, _it has no way to deserialize it_. ,  so instead, it stores the extension code as an "undefined" json.Node.

This is fairly safe when the given extension is only "used".

But for "required" extensions it means part of the data you might need to fully traverse the scene is "hidden" in these unknown json nodes.

> I've opended a issue regarding KHR_materials_variants on the blender project.

Could you point me to the issue? so I can track it too?

**Comment by BoyBaykiller** [comment]
https://projects.blender.org/blender/blender/issues/128647
